### PR TITLE
<numeric> Optimize gcd to use builtins

### LINKED ISSUE #298: <numeric>: Make gcd() faster with constexpr bitscan intrinsics
Our `gcd()` implementation contains the following workaround:

https://github.com/microsoft/STL/blob/58bb49d63d92e7a0346a05af29816aeea6b4cf0f/stl/inc/numeric#L847-L860

For #25 (WG21-P0553 `<bit>` Rotating And Counting Functions), we're getting compiler support for `constexpr` bitscans (with different intrinsic names). We should use this in `gcd()`.

Note that because `<bit>` is a C++20 header, we can't include it from `<numeric>`, nor can we use its non-`_Ugly` functions to implement C++17 features like `gcd()`. However, we can use the compiler intrinsics directly, or define `_Ugly` wrappers for them in centralized headers.

Also tracked by Microsoft-internal VSO-597250.

### PR DESCRIPTION
**Pull request by miscco**
Description
===========
This fixes #298 by moving the meat of the bitscan implementation to the limits header.

This seems unfortunate but is the smallest common denominator of `<numeric>` and `<bit>`

Also we cannot go without limits anyway because the implementation needs it.

Checklist
=========

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

**Comment by BillyONeal** [comment]
Test failures look legit on this one :)

**Comment by miscco** [comment]
Hm I compiled for me, I will check again :(

**Comment by miscco** [comment]
Go home test you are drunk:
>C:\agent\_work\1\a\x86\out\inc\limits(1021,26): note: candidate template ignored: requirement '_Is_standard_unsigned_integer<int>' was not satisfied [with _Ty = int]

**Comment by miscco** [comment]
The full error is:
>In file included from C:\agent\_work\1\s\llvm-project\libcxx\test\std\numerics\numeric.ops\numeric.ops.gcd\gcd.pass.cpp:16:
>C:\agent\_work\1\a\x86\out\inc\numeric(864,39): error: no matching function for call to '_Countl_zero'
>
>    return static_cast<unsigned long>(_Countl_zero(~_Mask));
>                                     ^~~~~~~~~~~~
>C:\agent\_work\1\a\x86\out\inc\numeric(898,39): note: in instantiation of function template specialization 'std::_Stl_bitscan_forward<unsigned char>' requested here
>
>    const auto _Mx_trailing_zeroes  = _Stl_bitscan_forward(_Mx_magnitude);
>
> C:\agent\_work\1\s\llvm-project\libcxx\test\std\numerics\numeric.ops\numeric.ops.gcd\gcd.pass.cpp(48,45): note: in instantiation of function template specialization 'std::gcd<signed char, signed char>' requested here
> 
>     assert(static_cast<Output>(out) == std::gcd(value1, value2));
>
> C:\agent\_work\1\s\llvm-project\libcxx\test\std\numerics\numeric.ops\numeric.ops.gcd\gcd.pass.cpp(64,27): note: in instantiation of function template specialization 'test0<signed char, signed char, signed char>' requested here
> 
>             accumulate &= test0<S1, S2, Output>(TC.x, TC.y, TC.expect);
> 
>                           ^
> 
> C:\agent\_work\1\s\llvm-project\libcxx\test\std\numerics\numeric.ops\numeric.ops.gcd\gcd.pass.cpp(100,19): note: in instantiation of function template specialization 'do_test<signed char, signed char>' requested here
> 
>     static_assert(do_test<signed char>(), "");
> 
>                   ^
> 
> C:\agent\_work\1\a\x86\out\inc\limits(1021,26): note: candidate template ignored: requirement '_Is_standard_unsigned_integer<int>' was not satisfied [with _Ty = int]
> 
> _NODISCARD constexpr int _Countl_zero(const _Ty _Val) noexcept {

So in the test the type is `signed_char` that is passed into `make_unsigned_t` to deduce `_Common_unsigned`. I guess that should make it a `unsigned char` which is the type of `_Mx_magnitude`.

So the type passed to `_Countl_zero` is `unsigned char` where does it get the `int` from?

**Comment by miscco** [comment]
It seems that for whatever reason the template argument deduction fails. if I specify `_Countl_zero<_Unsigned>` it works.

That however smells terribly like a compiler bug

**Comment by miscco** [comment]
So tests are good, although I would like an opinion whether this is a compiler bug or just me not understanding C++.

The only thing I could com up with is that `~_Mask` is changing the type but please tell me it doesnt

File: stl/inc/numeric [hunk 1]
```diff
@@ -859,7 +859,10 @@ _NODISCARD constexpr auto _Abs_u(const _Arithmetic _Val) noexcept {
 
 // FUNCTION TEMPLATE _Stl_bitscan_forward
 template <class _Unsigned>
-constexpr unsigned long _Stl_bitscan_forward(_Unsigned _Mask) noexcept {
+_NODISCARD constexpr unsigned long _Stl_bitscan_forward(_Unsigned _Mask) noexcept {
+#ifdef __cpp_lib_bitops // TRANSITION, VSO-1020212
+    return static_cast<unsigned long>(_Countl_zero<_Unsigned>(~_Mask));
```
Comment (miscco) [inline stl/inc/numeric:864]:
```suggestion
    return static_cast<unsigned long>(_Countl_zero(static_cast<_Unsigned>(~_Mask));

```

**Comment by AlexGuteniev** [comment]
> that `~_Mask` is changing the type

Why not?

https://eel.is/c++draft/expr.unary.op#10

>Integral promotions are performed. The type of the result is the type of the promoted operand.

https://eel.is/c++draft/conv.prom#2

>A prvalue of type char16_­t, char32_­t, or wchar_­t ([basic.fundamental]) can be converted to a prvalue of the first of the following types that can represent all the values of its underlying type: int, unsigned int, long int, unsigned long int, long long int, or unsigned long long int.

File: stl/inc/numeric [hunk 2]
```diff
@@ -859,7 +859,11 @@ _NODISCARD constexpr auto _Abs_u(const _Arithmetic _Val) noexcept {
 
 // FUNCTION TEMPLATE _Stl_bitscan_forward
 template <class _Unsigned>
-constexpr unsigned long _Stl_bitscan_forward(_Unsigned _Mask) noexcept {
+_NODISCARD constexpr unsigned long _Stl_bitscan_forward(_Unsigned _Mask) noexcept {
+#ifdef __cpp_lib_bitops // TRANSITION, VSO-1020212
+    return static_cast<unsigned long>(_Countl_zero(static_cast<_Unsigned>(~_Mask));
```
Comment (CaseyCarter) [inline stl/inc/numeric:864]:
```suggestion
    return static_cast<unsigned long>(_Countl_zero(static_cast<_Unsigned>(~_Mask)));
```

Comment (miscco) [reply stl/inc/numeric]: Oh my how did I not stage that change

Comment (CaseyCarter) [reply stl/inc/numeric]: You were probably still reeling from the madness of discovering that unary operators perform integral promotions. This is *not* my favorite feature of C++.

Comment (miscco) [reply stl/inc/numeric]: That was one of those moments where you just shut the machine down and go for a walk

File: stl/inc/numeric [hunk 3]
```diff
@@ -859,7 +859,11 @@ _NODISCARD constexpr auto _Abs_u(const _Arithmetic _Val) noexcept {
 
 // FUNCTION TEMPLATE _Stl_bitscan_forward
 template <class _Unsigned>
-constexpr unsigned long _Stl_bitscan_forward(_Unsigned _Mask) noexcept {
+_NODISCARD constexpr unsigned long _Stl_bitscan_forward(_Unsigned _Mask) noexcept {
+#ifdef __cpp_lib_bitops // TRANSITION, VSO-1020212
+    return static_cast<unsigned long>(_Countl_zero(static_cast<_Unsigned>(~_Mask)));
+
```
Comment (StephanTLavavej) [inline stl/inc/numeric:865]: I believe this empty line is unnecessary:

```suggestion
```

Comment (miscco) [reply stl/inc/numeric:865]: I was soo close

File: stl/inc/bit [hunk 4]
```diff
@@ -23,8 +23,6 @@ _STL_DISABLE_CLANG_WARNINGS
 #undef new
 
 _STD_BEGIN
-enum class endian { little = 0, big = 1, native = little };
```
Comment (CaseyCarter) [inline stl/inc/bit:26]: Drive-by: I put this in synopsis order (at the end) while I was here.

**Comment by CaseyCarter** [comment]
Thanks for the performance enhancement! 🐎
